Fix/jetty 12/10541/byte buffer accumulator

gregw opened the pull request.

gregw committed:
Review ByteBufferAccumulator #10541

 + Added `isRetained()` to `Retainable`
 + Made `Chunk` is-a `RetainableByteBuffer`
 + retain rather than copy chunks where possible

gregw committed:
Review ByteBufferAccumulator #10541

 + Added `isRetained()` to `Retainable`
 + Made `Chunk` is-a `RetainableByteBuffer`
 + retain rather than copy chunks where possible

gregw committed:
Review ByteBufferAccumulator #10541

 + Added `isRetained()` to `Retainable`
 + Made `Chunk` is-a `RetainableByteBuffer`
 + retain rather than copy chunks where possible

gregw committed:
Review ByteBufferAccumulator #10541

 + Added `isRetained()` to `Retainable`
 + Made `Chunk` is-a `RetainableByteBuffer`
 + retain rather than copy chunks where possible
 + Merged aggregator into accumulator
 + Deleted ByteBufferAggregator.java

gregw committed:
Review ByteBufferAccumulator #10541

 + Added `isRetained()` to `Retainable`
 + Made `Chunk` is-a `RetainableByteBuffer`
 + retain rather than copy chunks where possible
 + Merged aggregator into accumulator
 + Deleted ByteBufferAggregator.java

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12/10541/byteBufferAccumulator

gregw committed:
Review ByteBufferAccumulator #10541

 + Added `isRetained()` to `Retainable`
 + Made `Chunk` is-a `RetainableByteBuffer`
 + retain rather than copy chunks where possible
 + Merged aggregator into accumulator
 + Deleted ByteBufferAggregator.java

lachlan-roberts commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ByteBufferAccumulator.java:
```diff
@@ -105,6 +120,94 @@ public void copyBuffer(ByteBuffer source)
+        {
+            int prefix = buffer == null ? 0 : buffer.remaining();
+            int minSize = prefix + toCopy;
+            int allocSize = (int)Math.min(_maxLength, ceilToNextPowerOfTwo(minSize));
```
Should this be min of the remaining space `_maxLength - getLength()` instead of just `_maxLength`?

gregw replied:
@lachlan-roberts I think it already is, since the `toCopy` has already been limited by `space = _maxLength - getLength()` above and `minSize = prefix + toCopy`

lachlan-roberts reviewed (commented):
@gregw note that the `ensureBuffer` method can allow you to bypass the max size because you can write directly to the buffer.
 
Why do we use the `ceilToNextPowerOfTwo` method only in `aggregate()` but not in `ensureBuffer()`?

gregw commented:
> note that the `ensureBuffer` method can allow you to bypass the max size because you can write directly to the buffer.

@lachlan-roberts yeah, but given that API, there is not much we can do about that....   Hmmm maybe if there is a max size, we should return a slice from ensureBuffer that has the capacity limited by maxSize?   I'll try that.

gregw commented:
> > note that the `ensureBuffer` method can allow you to bypass the max size because you can write directly to the buffer.
> 
> @lachlan-roberts yeah, but given that API, there is not much we can do about that.... Hmmm maybe if there is a max size, we should return a slice from ensureBuffer that has the capacity limited by maxSize? I'll try that.

@lachlan-roberts nah that doesn't work because the slices position and limit are independent.     I can't see how it can be done.  If the allocated buffer is larger than maxLength then that length can be used by that API.
Why is that API public?    It is not used externally?   I think I will deprecate the public nature of it and only use a private version

gregw committed:
Review ByteBufferAccumulator #10541

 + Added `isRetained()` to `Retainable`
 + Made `Chunk` is-a `RetainableByteBuffer`
 + retain rather than copy chunks where possible
 + Merged aggregator into accumulator
 + Deleted ByteBufferAggregator.java

gregw committed:
Review ByteBufferAccumulator #10541

 + Added `isRetained()` to `Retainable`
 + Made `Chunk` is-a `RetainableByteBuffer`
 + retain rather than copy chunks where possible
 + Merged aggregator into accumulator
 + Deprecated ByteBufferAggregator.java
 + Deprecated ByteBufferCallbackAccumulator

gregw commented:
@sbordet @lachlan-roberts this is becoming a bit of a swiss army knife.  Perhaps it is better to keep separate classes for specific purposes?

gregw committed:
Review ByteBufferAccumulator #10541

 + Added `isRetained()` to `Retainable`
 + Made `Chunk` is-a `RetainableByteBuffer`
 + retain rather than copy chunks where possible
 + Merged aggregator into accumulator
 + Deprecated ByteBufferAggregator.java
 + Deprecated ByteBufferCallbackAccumulator

gregw commented:
@sbordet @lachlan-roberts @lorban here is an alternate proposal rather than making a one-size-fits all aggregator.

Instead we enhance the RetainableByteBuffer API.      Firstly we add some default methods for appending content:
 + `int append(byte[] buffer, int offset, int length)`
 + `void append(ByteBuffer buffer)`
 + `void append(ByteBuffer buffer, Runnable releaser)`
 + `void append(RetainableByteBuffer buffer)`

The default implementations will append as much as they can given the capacity of the return from  `getByteBuffer()`. We also add some utility methods with default implementations for:
+ `void writeTo(ByteBuffer buffer)`
+ `void writeTo(OutputStream out) throws IOException`
+ `void writeTo(Content.Sink sink, Callback callback)`

Then we add some static utility types as subtypes of the `RetainableByteBuffer` interface obtained via statics:
 + `RetainableByteBuffer newGrowable(ByteBufferPool pool, int initialSize, int maxSize)`
 + `RetainableByteBuffer newAccumulator(ByteBuffer pool, long maxLength)`

Ultimately all the different accumulators, aggregators etc. live behind the `RetainableByteBuffer` API.   So we don't allocate an accumulator from which we can take a `RetainableByteBuffer`, instead we just use a `RetainableByteBuffer` that can internally accumulate other buffers.  Likewise, we don't create and aggregator that can grow an internal buffer that we ultimately take, we just have a `RetainableByteBuffer` that can grow.

Another way of looking at this, is that we can move a lot of t he methods from `BufferUtil` to `RetainableByteBuffer` and fix up the legacy inconsistencies and poor naming at the same time!

lorban commented:
I've wondered too if we should have read/write methods on `RetainableByteBuffer` and stop using `ByteBuffer` in our APIs entirely, especially since we already have an incomplete set of methods working on the byte buffer: `clear`, `hasRemaining`... and I agree that it would probably be a good idea.

But I am not convinced about making `RetainableByteBuffer` more than a `ByteBuffer` wrapper. The biggest problem IMHO would be the `getByteBuffer()` method of an `AccumulatingRetainableByteBuffer`, but I also feel that this change would cram too much delicate logic into a single class, risking many subtle bugs.

I think the accumulation/aggregation concept should be kept separate from the existing single buffer concept, but we could introduce a new interface that would standardize the API and the lifecycling of the buffers:

```java
public interface RetainableMemory
{
  // returns int as maybe not all bytes will be appended as a RetainableMemory can have a size limit
  int append(RetainableByteBuffer buffer);

  // returns void as this is non-blocking, so a way to figure out how much data was not written is needed
  void writeTo(Content.Sink sink, Callback callback);

  // returns long as there's potentially more than 1 buffer, i.e.: > 2GB
  long remaining();
}
```

This highlights some problems, like:
 - Should `writeTo` release the buffers when it succeeds?
 - Should `writeTo` release the buffers when it fails? Including the ones it did not write?
and I'm sure there are others I did not think of.

gregw commented:
@lorban why do you see getByteBuffer as a problem? The accumulators and aggregators already have equivalent method to get the combined content as a single buffer, so it is no different. Of course it would be better if writeTo style methods could be used to avoid that copy.

I'll give this a go in a different branch to see what it looks like....

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12/10541/byteBufferAccumulator

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12/10541/byteBufferAccumulator

gregw commented:
@lorban 
> This highlights some problems, like:
> 
> * Should `writeTo` release the buffers when it succeeds?
> * Should `writeTo` release the buffers when it fails? Including the ones it did not write?
>   and I'm sure there are others I did not think of.

No and No.  A `RetainableByteBuffer` is released when `release()` is called.  Same applies for aggregates and accumulators.     

In the case of a `writeTo(Sink, boolean, Callback)`, there is already a callback to indicate success or failure.
